### ISSUE TITLE
[studio] MongoDB and MariaDB errors showing up when upgrading from 3.1.x to 4.1.x/4.2.x

### ISSUE THREAD

**Issue by Euquimides** [craftercms/craftercms#6308]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] The issue is in the develop branch
- [X] The issue is in the latest released 4.1.x
- [ ] The issue is in the latest released 4.0.x
- [ ] The issue is in the latest released 3.1.x

### Describe the issue

Upgrading from 3.1.x -> 4.1.x/4.2.x is showing errors related with MongoDB and MariaDB. An error with MongoDB will show up when the user selects to backup their data, while an error with MariaDB version will show up when not. 

### Steps to reproduce

Steps:

1. Run the upgrade-search.sh script for updating indexes from ElasticSearch to OpenSearch.
2. Then run the upgrade-target.sh script.
3. Pick either to backup or not your data
4. See the issue

### Relevant log output

```shell
If the user selects to backup their data, the following error will show up:

Starting MongoDB
------------------------------------------------------------------------
about to fork child process, waiting until server is ready for connections.
forked process: 249536
ERROR: child process failed, exited with 100
To see additional information in this output, start without the "--fork" option.
------------------------------------------------------------------------
Backing up mongodb
------------------------------------------------------------------------
2023-12-04T18:02:37.768-0600	Failed: can't create session: failed to connect to mongodb://localhost:27020/: server selection error: server selection timeout, current topology: { Type: Single, Servers: [{ Addr: localhost:27020, Type: Unknown, Last error: connection() error occurred during connection handshake: dial tcp 127.0.0.1:27020: connect: connection refused }, ] }
Unable to continue, an error occurred or the script was forcefully stopped
Caught: java.lang.RuntimeException: Process '[./crafter.sh, backup]' exited with non-successful value 1
java.lang.RuntimeException: Process '[./crafter.sh, backup]' exited with non-successful value 1
	at utils.ScriptUtils.executeCommand(ScriptUtils.groovy:78)
	at utils.ScriptUtils.executeCommand(ScriptUtils.groovy:55)
	at upgrade.upgrade-target.backupData(upgrade-target.groovy:112)
	at upgrade.upgrade-target.upgrade(upgrade-target.groovy:514)
	at upgrade.upgrade-target.run(upgrade-target.groovy:545)

And when selecting not to backup data, the upgrade will throw the following error:

Caught: java.nio.file.FileSystemException: /home/juanvasquez/crafter/bundles/crafterbundle3130EE/crafter/bin/mariaDB4j-app.jar/mariaDB4j-app-2.6.0.9.jar: Not a directory
java.nio.file.FileSystemException: /home/juanvasquez/crafter/bundles/crafterbundle3130EE/crafter/bin/mariaDB4j-app.jar/mariaDB4j-app-2.6.0.9.jar: Not a directory
	at java.base/jdk.internal.reflect.GeneratedMethodAccessor19.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at upgrade.upgrade-target.syncFile(upgrade-target.groovy:342)
	at jdk.internal.reflect.GeneratedMethodAccessor18.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at upgrade.upgrade-target$_doUpgrade_closure9$_closure13.doCall(upgrade-target.groovy:483)
	at jdk.internal.reflect.GeneratedMethodAccessor17.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at upgrade.upgrade-target$_doUpgrade_closure9.doCall(upgrade-target.groovy:481)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at upgrade.upgrade-target.doUpgrade(upgrade-target.groovy:479)
	at upgrade.upgrade-target$doUpgrade$4.callCurrent(Unknown Source)
	at upgrade.upgrade-target.upgrade(upgrade-target.groovy:515)
	at upgrade.upgrade-target$upgrade$0.callCurrent(Unknown Source)
	at upgrade.upgrade-target.run(upgrade-target.groovy:544)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

**Comment by phuongnq** [comment]
I sent PR to fix the MariaDB4j issue. The location of the file `mariaDB4j-app.jar` was incorrect as it was copied as a folder.

Regarding the MongoDB startup while upgrading, I cannot reproduce the issue. 
@Euquimides While running the `upgrade-target.sh` script, I believe you have to provide a path of which bundle to be updated. Ex:

```
./upgrade-target.sh /home/pnguyen/Downloads/tmp/crafter-cms-authoring-3.1.31/crafter
```

If your new bundle has MongoDB (Profile & Social modules), the bundle to be upgraded also needs to have the MongoDB (In this case `/home/pnguyen/Downloads/tmp/crafter-cms-authoring-3.1.31/crafter` should include MongoDB) to start the MongoDB server. Please verify if your bundles both include MongoDB module.

**Comment by sumerjabri** [comment]
@phuongnq I think I may have made the same mistake in 4.0. Please send PRs to `4.0.x` and `develop`

**Comment by phuongnq** [comment]
@sumerjabri @Euquimides Since the version of MongoDB in 4.2.x (MongoDB version 6) is different from the version in 3.1.x (MongoDB version is 3.4.4), we need to manually update MongoDB.

Please refer to the following steps as an example while I upgrade the `/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter` to v4.2.0:

1. Perform the upgrade-target.sh with `MONGODB_HOME`:

```
MONGODB_HOME=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/mongodb ./upgrade-target.sh /home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter
```
Note that I have to specify the `MONGODB_HOME` so that the backup script will run with the `mongod` version `3.4.4` in the old bundle

2. Manually upgrade MongoDB from 3.4.4 to v6:

Refer to https://www.mongodb.com/docs/manual/release-notes/6.0-upgrade-standalone/

We need to upgrade from each release version and run `setFeatureCompatibilityVersion` to DB.

Here are the download links for Ubuntu to download each version:

http://downloads.mongodb.org/linux/mongodb-linux-x86_64-3.6.0.tgz
http://downloads.mongodb.org/linux/mongodb-linux-x86_64-4.0.0.tgz
https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu1804-4.2.25.tgz
https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2004-4.4.26.tgz
https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2004-5.0.23.tgz
https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2204-6.0.12.tgz


Run `mongod` from each release version and set the compatibility version:

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-3.6.0/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "3.6" } )
```

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-4.0.0/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "4.0" } )
```

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu1804-4.2.25/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "4.2" } )
```

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2004-4.4.26/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "4.4" } )
```

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2004-5.0.23/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "5.0" } )
```

```
/home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2204-6.0.12/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
```

Run the setFeatureCompatibilityVersion command against the admin database:

```
db.adminCommand( { setFeatureCompatibilityVersion: "6.0" } )
```

3. Run the post-upgrade script

```
/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/upgrade/post-upgrade.sh
```

**Comment by phuongnq** [comment]
I also sent a PR to fix the post-upgrade scripts. Please re-test after the PR is merged.

**Comment by sumerjabri** [comment]
@alhambrav please update the 4.2 upgrade instructions for Social/Profile to include the steps detailed above for MongoDB upgrade.

**Comment by Euquimides** [comment]
> @sumerjabri @Euquimides Since the version of MongoDB in 4.2.x (MongoDB version 6) is different from the version in 3.1.x (MongoDB version is 3.4.4), we need to manually update MongoDB.
> 
> Please refer to the following steps as an example while I upgrade the `/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter` to v4.2.0:
> 
>     1. Perform the upgrade-target.sh with `MONGODB_HOME`:
> 
> 
> ```
> MONGODB_HOME=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/mongodb ./upgrade-target.sh /home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter
> ```
> 
> Note that I have to specify the `MONGODB_HOME` so that the backup script will run with the `mongod` version `3.4.4` in the old bundle
> 
>     2. Manually upgrade MongoDB from 3.4.4 to v6:
> 
> 
> Refer to https://www.mongodb.com/docs/manual/release-notes/6.0-upgrade-standalone/
> 
> We need to upgrade from each release version and run `setFeatureCompatibilityVersion` to DB.
> 
> Here are the download links for Ubuntu to download each version:
> 
> http://downloads.mongodb.org/linux/mongodb-linux-x86_64-3.6.0.tgz http://downloads.mongodb.org/linux/mongodb-linux-x86_64-4.0.0.tgz https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu1804-4.2.25.tgz https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2004-4.4.26.tgz https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2004-5.0.23.tgz https://fastdl.mongodb.org/linux/mongodb-linux-x86_64-ubuntu2204-6.0.12.tgz
> 
> Run `mongod` from each release version and set the compatibility version:
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-3.6.0/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "3.6" } )
> ```
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-4.0.0/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "4.0" } )
> ```
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu1804-4.2.25/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "4.2" } )
> ```
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2004-4.4.26/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "4.4" } )
> ```
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2004-5.0.23/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "5.0" } )
> ```
> 
> ```
> /home/pnguyen/Downloads/mongodb-linux-x86_64-ubuntu2204-6.0.12/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020
> ```
> 
> Run the setFeatureCompatibilityVersion command against the admin database:
> 
> ```
> db.adminCommand( { setFeatureCompatibilityVersion: "6.0" } )
> ```
> 
>     3. Run the post-upgrade script
> 
> 
> ```
> /home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/upgrade/post-upgrade.sh
> ```

Hi @phuongnq ! The command `db.adminCommand( { setFeatureCompatibilityVersion: "3.6" } )` for example, can be run or should be run directly with `mongosh` inside `CRAFTER_HOME/crafter_authoring/bin/mongodb` through the port specified when running `mongod`, correct? Is that in the instructions it is not so clear, and it will help for documentation purposes. For example:

Run mongod from each release version and set the compatibility version:

```/home/pnguyen/Downloads/mongodb-linux-x86_64-3.6.0/bin/mongod --dbpath=/home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/data/mongodb --directoryperdb --journal --port 27020```

Go to `home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/mongodb/bin` and run the command:

```./mongosh --port 27020```

This to connect to the current mongodb instance.

Run the setFeatureCompatibilityVersion command against the admin database:

```db.adminCommand( { setFeatureCompatibilityVersion: "3.6" } )```

 Are this list of steps correct? Thank you for your answer and specification!

**Comment by phuongnq** [comment]
Hi @Euquimides 
Yes, your steps are correct. Any MongoDB clients are ok and we can use `mongosh` for this. Just a note while using `mongosh`. We can specify a URL to the connection string to specify the `admin` database with

```
./mongosh mongodb://localhost:27020/admin
```

I'm not testing the `mongosh` version with each MongoDB server version but hopefully, the latest (in `home/pnguyen/Downloads/crafter-cms-authoring-3.1.32-SNAPSHOT/crafter/bin/mongodb/bin` after upgraded) will work with all the server versions (Please help to verify this!). If not we may need to download each specific `mongosh` version for each upgrade.

**Comment by Euquimides** [comment]
Hi! Verified fix for upgrade from v3.1.x -> v4.2. Also run `mongosh` against all server versions, its working fine now. Thank you!

**Comment by alhambrav** [comment]
https://docs.craftercms.org/en/4.1/by-role/system-admin/upgrade/index.html#upgrading-mongodb

https://docs.craftercms.org/en/4.0/system-administrators/upgrade/index.html#mongodb-upgrade
